scrapy/dispatch/dispatcher.py

```diff
@@ -0,0 +1,386 @@
+        """
+        assert callable(receiver), "Signal receivers must be callable."
+        # Check for **kwargs
+        if not func_accepts_kwargs(receiver):
+           warnings.warn("The use of handlers that don't accept "
+                          "**kwargs has been deprecated, plese refer "
+                          "to the Signals API documentation.",
+                          ScrapyDeprecationWarning, stacklevel=3)
+           self.receiver_accepts_kwargs[_make_id(receiver)] = False
+        else:
+            self.receiver_accepts_kwargs[_make_id(receiver)] = True
+
+        if dispatch_uid:
+            lookup_key = (dispatch_uid, _make_id(sender))
+        else:
+            lookup_key = (_make_id(receiver), _make_id(sender))
+
+        if weak:
+            ref = weakref.ref
+            receiver_object = receiver
+            # Check for bound methods
+            if hasattr(receiver, '__self__') and hasattr(receiver, '__func__'):
+                ref = WeakMethod
+                receiver_object = receiver.__self__
+            if six.PY3 and sys.version_info >= (3, 4):
+                receiver = ref(receiver)
+                weakref.finalize(receiver_object, self._remove_receiver)
+            else:
+                receiver = ref(receiver, self._remove_receiver)
+
+        with self.lock:
+            self._clear_dead_receivers()
+            for r_key, _ in self.receivers:
+                if r_key == lookup_key:
+                    break
+            else:
+                self.receivers.append((lookup_key, receiver))
+            self.sender_receivers_cache.clear()
+
+    def disconnect(self, receiver=None, sender=None, dispatch_uid=None):
+        """
+        Disconnect receiver from sender for signal.
+
+        If weak references are used, disconnect need not be called. The receiver
+        will be remove from dispatch automatically.
+
+        Arguments:
+
+            receiver
+                The registered receiver to disconnect. May be none if
+                dispatch_uid is specified.
+
+            sender
+                The registered sender to disconnect
+
+            dispatch_uid
+                the unique identifier of the receiver to disconnect
+        """
+        if dispatch_uid:
+            lookup_key = (dispatch_uid, _make_id(sender))
+        else:
+            lookup_key = (_make_id(receiver), _make_id(sender))
+
+        disconnected = False
+        with self.lock:
+            self._clear_dead_receivers()
+            for index in range(len(self.receivers)):
+                (r_key, _) = self.receivers[index]
+                if r_key == lookup_key:
+                    disconnected = True
+                    del self.receivers[index]
+                    break
+            self.sender_receivers_cache.clear()
+            if disconnected:
+                if _make_id(receiver) in self.receiver_accepts_kwargs:
+                    del self.receiver_accepts_kwargs[_make_id(receiver)]
```

Comment:
There's a race condition here, you can use `self.receiver_accepts_kwargs.pop(_make_id(receiver), None)` to get rid of it